### ISSUE TITLE
Replace RTTI by LLVM-style RTTI?

### ISSUE THREAD

**Issue by Keno** [AliveToolkit/alive2#915]
In https://github.com/JuliaPackaging/Yggdrasil/pull/4271, I managed to build Alive for the Julia binary distribution system.
However, as is, I cannot use the LLVM plugin for translation validation within Julia [1], because Julia's LLVM ABI is no-RTTI, no EH.
Additionally, as I found out in that attempt, it's currently not possible to build an RTTI-enabled LLVM for Windows (at least with the backend config we have) due to a separate LLVM issue (https://reviews.llvm.org/D109192). With that in mind,

1. I was wondering if you'd be amenable to replacing Alive2's use of RTTI by an LLVM-style RTTI implementation.
2. For EH, are there any places where Alive2 actually expects to unwind *through* stack frames that are created by LLVM? If not, can the requirement for LLVM to be built with EH be relaxed?

Thanks!

[1] There's additional, semantic, things that are breaking this too, right now, but I'm trying to get through the plumbing first.

**Comment by nunoplopes** [comment]
I think the exceptions we have are just within our code. Z3 doesn't throw, nor does LLVM. LLVM doesn't call back into our code, so all fine.
The requirement for exceptions was just to make cmake hacking easier (we inherit LLVM flags that contain the EH disabling flags).

Regarding RTTI, I'm certainly not going to do that work. If you are willing to do it, I would merge a PR, as long as it's not a significant burden on development.
An alternative would be to hoist all LLVM interfacing code to separate files so they are compiled without RTTI. They then export the functionality to the rest of Alive2. Our interaction with LLVM is fairly limited.

**Comment by Keno** [comment]
Sounds good, I'll look at both options and see which is easier.

**Comment by regehr** [comment]
or just use the LLVM C API? I've never actually use it but I do often look at its API while hacking on LLVM just because it makes a lot of things explicit that are implicit in C++, and it seems totally fine

**Comment by nunoplopes** [comment]
If it supports everything we need, it could be an option. But I'm not sure we can extract all intrinsics and metadata through the C API.

**Comment by ryan-berger** [comment]
Thought I'd chime in.

I wrote Alive bindings for Go, which you can see here: https://github.com/ryan-berger/golang_tv

For the Go version I used the C API to build out the LLVM IR in memory, and then I wrote a small C wrapper around Alive that I used to invoke Alive. The LLVM C API just uses a C wrapper around the LLVM C++ API so once you start passing pointers around it is essentially the same thing. 

There's plenty of ways to add metadata such as attributes in the C API, which is taken care of for me by the lovely folks over at TinyGo (a LLVM-based Go compiler)

TLDR, the C API works very well with Alive, and can do just about everything the C++  API can do. Would recommend writing a small C wrapper for Alive's API

**Comment by nunoplopes** [comment]
closing this, but feel free to open a PR if you manage to make it work.
